### DISCUSSION TITLE
NavigationBar Control

### DISCUSSION THREAD

**Discussion by kazo0** [unoplatform/uno.toolkit.ui#45] [Ideas]
We need a control that, on iOS and Android, will hook into the Android Toolbar and iOS UINavigationBar and provide a way to customize the native navigation bar. On other platforms, this control will simply be a façade for the `CommandBar`. The logic to forward DependencyProperties to the native Toolbar/UINavigationBar already exists as part of the `CommandBarRenderer` for [iOS](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Controls/CommandBarRenderer.iOS.cs) and [Android](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Controls/CommandBarRenderer.Android.cs). This logic will be imported to the Toolkit and will be renamed to `NavigationBarRenderer`. Below is the proposed contract for this new NavigationBar control. The properties outlined here will replace the current [Extensions](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/Controls/CommandBarRenderer.Android.cs) that are being used as part of the [`CommandBar`](https://platform.uno/docs/articles/controls/CommandBar.html) as well as having similar DependencyProperties to those of `CommandBar`, which will be forwarded to the inner `CommandBar` that the `NavigationBar` is composed of. 

|       Property       |                Type               |                  Comments                 |
|:--------------------:|:---------------------------------:|:-----------------------------------------:|
|   NavigationButton   |           `AppBarButton`          |                                           |
| NavigationButtonMode |       `NavigationButtonMode`      | Defaults to `NavigationButtonMode.Native` |
|       Subtitle       |              `string`             |                Android Only               |
|    PrimaryCommands   | `IObservableVector<AppBarButton>` |                                           |
|   SecondaryCommands  | `IObservableVector<AppBarButton>` |           Not available for iOS           |
|        Content       |              `object`             |                                           |

We will apply the refactoring of the CommandBar Extensions from this [discussion](https://github.com/unoplatform/uno/discussions/6057#discussioncomment-781583) to the `NavigationBar`. 

On Android and iOS, the NavigationBar will not be rendered where it is placed in the XAML. It will serve as a placeholder from which to read the information we need to pass to the `NavigationBarRenderer` that will, in turn, setup the native Android Toolbar or iOS UINavigationBar. On other platforms, the default NavigationBar control template will simply contain a `CommandBar` whose properties are template bound to the same properties on the `CommandBar`. So, on platforms like WASM/Skia, the `PrimaryCommands`, `SecondaryCommands`, `Content`, etc. will be forwarded to the inner `CommandBar`

**Comment by kazo0** [comment]
Based on https://platform.uno/docs/articles/controls/CommandBar.html#appbarbutton, we are only allowing `AppBarButton` as an acceptable Primary/Secondary Command

**Comment by davidjohnoliver** [comment]
What's the motivation for creating this new control, versus continuing to have this functionality in CommandBar?

**Reply by kazo0** [reply]
It's confusing that this has ever been part of CommandBar at all. Now that we are importing the CommandBar/AppBar code from WinUI, we can have CommandBar work the same across all platforms and separate out the logic for a native navigation bar. The other option would be to keep all the current CommandBar code with the renderer in the WUXC namespace, and the new CommandBar would exist in the MUXC namespace. So we'd have two separate CommandBar implementations which would probably cause even more confusion

**Reply by davidjohnoliver** [reply]
Also just to make sure, it's not actually called NavBar, right? `NavigationBar`, yes?

**Reply by kazo0** [reply]
Yeah that makes more sense :)

**Reply by davidjohnoliver** [reply]
Agreed that the present CommandBar situation is confusing for app developers in multiple ways. 

At some point it was under consideration to associate native navigation bar functionality with the Page.TopAppBar property, presumably this is a different direction? Are there shortcomings with the TopAppBar approach?

At present, the native modes of CommandBar, Frame, and Page are all intercoupled. Is that going to pose problems for lifting the CommandBar native behavior out of Uno.UI into an external library?

**Reply by francoistanguay** [reply]
Dont think property names should contain Navigation or NavigationBar as it's redundant.

Also, can we toggle between "Modes" by switching presenters and removing the Mode property somehow?

Should the LHS button also be called some type of Command like "MainCommand" or something?

Can Content be anything of just some string Text? 

Could SubTitle be anything as well? 

I thought there was a Titlle extension property at some point?

**Reply by kazo0** [reply]
@davidjohnoliver The TopAppBar approach was something I noticed as a possibility as well. I think it is still an option to do it this way. The AppBar seems to be able to be aware of whether it is part of the BottomAppBar/TopAppBar as opposed to just being an "inline" AppBar. I figured we could start with one option and if it gets too hairy we can explore the other option

**Comment by kazo0** [comment]
|      Property     |                Type               |        Comments       |
|:-----------------:|:---------------------------------:|:---------------------:|
|    MainCommand    |           `AppBarButton`          |                       |
|      Subtitle     |              `string`             |      Android Only     |
|  PrimaryCommands  | `IObservableVector<AppBarButton>` |                       |
| SecondaryCommands | `IObservableVector<AppBarButton>` | Not available for iOS |
|      Content      |              `string` or `FrameworkElement`             |                       |

In this revision:

- `NavigationButton` is renamed to `MainCommand`
- NavigationButtonMode is removed. I believe that, within the `NavigationBarRenderer`, we can check whether the `MainCommand`'s `Command` Property is set and, if so, we can then assume that we need to place the AppBarButton in the native UINavigationBar. The main use case for the old [NavigationCommand](https://platform.uno/docs/articles/controls/CommandBar.html#navigationcommand) was to override the action of the native left-side button to provide logic for things like popping up a dialog before navigating. (Need @jeanplevesque and/or @MatFillion to review). Instead of checking whether the `AppBarButton` exists or is visible [here](https://github.com/unoplatform/uno/blob/338189fa488e21f8049a83aa41f35d33dd505626/src/Uno.UI/Controls/CommandBarNavigationItemRenderer.iOS.cs#L103-L109), we can simply check if the `AppBarButton`'s `Command` property is set, and, if so, we know we are overriding the native behavior and therefore we can set `native.LeftBarButtonItem`

**Reply by jeanplevesque** [reply]
For custom actions on back navigation, you can check my proposition here:
https://github.com/unoplatform/uno/issues/4580
